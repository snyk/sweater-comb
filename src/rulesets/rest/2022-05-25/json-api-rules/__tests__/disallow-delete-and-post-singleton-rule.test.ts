import { OpenAPIV3 } from "@useoptic/openapi-utilities";
import { RuleRunner, TestHelpers } from "@useoptic/rulesets-base";
import { context } from "../../__tests__/fixtures";
import { doNotAllowDeleteOrPostIdForSingleton } from "../disallow-delete-and-post-singleton-rule";

const baseJson = TestHelpers.createEmptySpec();

describe("singleton disallowed methods", () => {
  test.each([["post"], ["delete"]])(
    "fails when adding %s method to singleton",
    (method) => {
      const afterJson = {
        ...baseJson,
        paths: {
          "/api/example": {
            "x-snyk-resource-singleton": true,
            [method]: {
              responses: {
                "200": {
                  description: "",
                  content: {
                    "application/json": {
                      schema: {
                        type: "object",
                        properties: {},
                      },
                    },
                  },
                },
              },
            },
          },
        },
      } as OpenAPIV3.Document;

      const ruleRunner = new RuleRunner([doNotAllowDeleteOrPostIdForSingleton]);
      const ruleInputs = {
        ...TestHelpers.createRuleInputs(baseJson, afterJson),
        context,
      };
      const results = ruleRunner.runRulesWithFacts(ruleInputs);
      expect(results.length).toBeGreaterThan(0);
      expect(results.every((result) => result.passed)).toBe(false);
      expect(results).toMatchSnapshot();
    },
  );
});
