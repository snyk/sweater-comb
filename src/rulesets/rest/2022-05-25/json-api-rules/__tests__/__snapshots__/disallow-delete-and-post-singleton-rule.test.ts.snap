// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`singleton disallowed methods fails when adding delete method to singleton 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {
          "method": "delete",
          "path": "/api/example",
        },
        "conceptualPath": [
          "operations",
          "/api/example",
          "delete",
        ],
        "jsonPath": "/paths/~1api~1example/delete",
        "kind": "operation",
      },
      "value": {
        "method": "delete",
        "pathPattern": "/api/example",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "delete is not allowed in JSON:API singletons",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "disallow singletons for delete or post",
    "passed": false,
    "received": undefined,
    "type": "requirement",
    "where": "DELETE /api/example",
  },
]
`;

exports[`singleton disallowed methods fails when adding post method to singleton 1`] = `
[
  {
    "change": {
      "location": {
        "conceptualLocation": {
          "method": "post",
          "path": "/api/example",
        },
        "conceptualPath": [
          "operations",
          "/api/example",
          "post",
        ],
        "jsonPath": "/paths/~1api~1example/post",
        "kind": "operation",
      },
      "value": {
        "method": "post",
        "pathPattern": "/api/example",
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "post is not allowed in JSON:API singletons",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "disallow singletons for delete or post",
    "passed": false,
    "received": undefined,
    "type": "requirement",
    "where": "POST /api/example",
  },
]
`;
